kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,238 @@
+/// to form a linear storage of bytes.
+///
+/// There device is split into blocks in two ways, into:
+/// - prepare blocks, which are the smallest unit of space
+/// that can be prepared for writing (see `BlockStorage::prepare_write`)
+/// - write blocks, which are the smallest unit of space that can be written
+///
+/// Every byte on the device belongs to exactly one prepare block,
+/// and to exactly one write block at the same time.
+///
+/// `W`: The size of a write block in bytes.
+/// `P`: The size of a prepare block in bytes.
+pub trait BlockStorage<const W: usize, const P: usize> {
+    /// Read data from a block, and into the buffer.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `Client::read_complete` once.
+    fn read(
+        &self,
+        region: &BlockIndex<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Write data from a buffer to storage.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// starting at the chosen `start_block`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT prepare the block for writing first.
+    /// The user of this function SHOULD ensure that a block is prepared
+    /// before calling `write` (see `prepare_write`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it SHOULD NOT be written again until it's prepared
+    /// as part of a prepare block.
+    /// Multiple writes to the same block are possible in this trait,
+    /// but the result is dependent on the nature of the underlying hardware,
+    /// and this interface does not provide a way to discover that.
+    ///
+    /// **Note** about raw flash devices: writes can turn bits from `1` to `0`.
+    /// To change a bit from `0` to `1`, a region must be erased (prepared).
+    ///
+    /// **Note**: some raw flash hardware only allows a limited number of writes before
+    /// an erase. If that is the case, the implementation MUST return an error
+    /// `ErrorCode::NOMEM` when this happens, even if the hardware silently
+    /// ignores the write.
+    ///
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `Client::write_complete` once.
+    fn write(
+        &self,
+        region: &BlockIndex<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Makes a region ready for writing.
+    ///
+    /// This corresponds roughly to the erase operation on raw flash devices,
+    /// but may also do nothing on devices where erasure is not necessary.
```

Comment:
This suggests that this API does not support the possibility of multiple writes (e.g., on flash)?

I think that's OK, but this should be explicit.

Reply:
I think you're reading this right: regions on flash are not ready for writing unless they get erased first. Without this method, we can't support flash.

This particular passage says that if you have 3D Xpoint memory or a plain old HDD, this may be implemented as a no-op.

Reply:
This starts to get a little weird. It means that a client on top of an HDD can read discarded data, while a client on top of flash can't. This then exposes which one the client is on top of, such that a client sitting on top of a HDD will do things that then won't work if you replace the underlying block store. This means that API is no longer transparent and isn't an interchangeable API.

Reply:
Perhaps I didn't make it clear, but my intention here is that the result of reads of discarded blocks are undefined. The note that discard may be a no-op is a hint to the driver author, rather than an API guarantee.

By assuming SSD semantics of required discard, the API remains interchangeable (if SSD/HDD interchangeability is what you refer to).

Would it be clearer with the following wording?

> This corresponds roughly to the erase operation on raw flash devices.
> An implementation may internally decide to do nothing when irrelevant,
> e.g. on devices where erasure is not necessary.

Reply:
The issue with "undefined" is that a client which assumes a particular implementation is then bound to it; it's no longer an interchangeable interface. The client uses a `BlockStorage`, but it has to be a specific one or it won't work. I.e., a client built on top of an HDD or NVRAM device could read after discarding, but then if you put it on top of an SSD it won't work.

Reply:
> This corresponds roughly to the erase operation on raw flash devices.
> An implementation may internally decide to do nothing when irrelevant,
> e.g. on devices where erasure is not necessary.

The issue is what the client will do.

Reply:
I see "undefined" more as "forbidden, but not enforced", as opposed to "implementation-defined". If a client decides to rely on something that's outside of the API contract, then it's the client's problem. Within the constraint that a client only uses allowed operations, it won't be able to distinguish between a HDD or SDD.

I admit it's not perfect to leave an undefined operation, but I think this is a local optimum for the useablity of this API. If we tried to eliminate that undefinedness, I expect we'd have to come up with something a lot more complicated. Perhaps we want to in the future, but I think this proposal is a good stepping stone to open up the familiar block storage functionality at all.